jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java

```diff
@@ -369,7 +369,7 @@ public void onFillable()
         HttpConnection last = setCurrentConnection(this);
         try
         {
-            while (getEndPoint().isOpen())
+            while (true)
```

Comment:
This was wrong, because by looking past the request (as a Content.Source) to the endpoint, this loop might become a noop, even when the endpoint has been consumed and buffered.    The end of reading should only be indicated by reading -1

Reply:
That should be stated in a comment.